File: leetcode/sum/259. 3Sum Smaller/js/main.js
// Time O(N^2)
const threeSumSmaller = (nums, target) => {
  let n = nums.length;
  let count = 0;

  nums.sort((a, b) => a - b);

  for (let i = 0; i < n - 2; i++) {
    if (nums[i] + nums[i + 1] + nums[i + 2] > target) break;

    let low = i + 1;
    let high = n - 1;

    // Находим все перестановки связанные с nums[i]
    while (low < high) {
      let sum = nums[i] + nums[low] + nums[high];

      if (sum < target) {
        // сколько элементов аналогично меньше target в промежутке от low до high ? это == high - low
        count += high - low;
        low++;
      } else {
        high--;
      }
    }
  }

  return count;
};

File: leetcode/sum/259. 3Sum Smaller/py/main.py
class Solution:
    # Time O(N^2)
    # Space O(N)
    def threeSumSmaller(self, nums, target):
        if not nums: return 0

        n = len(nums)
        ans = 0

        nums.sort()

        for i in range(n - 2):
            lo, hi = i + 1, n - 1

            while lo < hi:
                cur = nums[i] + nums[lo] + nums[hi]

                if cur >= target:
                    hi -= 1
                else:
                    ans += hi - lo
                    lo += 1

        return ans